api/serializers.py

```diff
@@ -47,37 +51,45 @@ def validate(self, data):
         return data
 
 
-def feed_type_validation(feed_type):
-    feed_choices = ["log4j", "cowrie", "all"]
-    generalHoneypots = GeneralHoneypot.objects.all().filter(active=True)
-    feed_choices.extend([hp.name.lower() for hp in generalHoneypots])  # FEEDS
-
-    if feed_type not in feed_choices:
-        logger.info(f"Feed type {feed_type} not in feed_choices {feed_choices}")
-        raise serializers.ValidationError("Invalid feed_type")
-    return feed_type
-
-
-class FeedsSerializer(serializers.Serializer):
-    feed_type = serializers.CharField(max_length=120)
-    attack_type = serializers.ChoiceField(choices=["scanner", "payload_request", "all"])
-    age = serializers.ChoiceField(choices=["persistent", "recent"])
-    format = serializers.ChoiceField(choices=["csv", "json", "txt"], default="json")
-
-    def validate_feed_type(self, feed_type):
-        logger.debug(f"FeedsSerializer - Validation feed_type: '{feed_type}'")
-        return feed_type_validation(feed_type)
-
-
-class FeedsResponseSerializer(serializers.Serializer):
-    feed_type = serializers.CharField(max_length=120)
-    value = serializers.CharField(max_length=120)
-    scanner = serializers.BooleanField()
-    payload_request = serializers.BooleanField()
-    first_seen = serializers.DateField(format="%Y-%m-%d")
-    last_seen = serializers.DateField(format="%Y-%m-%d")
-    times_seen = serializers.IntegerField()
-
-    def validate_feed_type(self, feed_type):
-        logger.debug(f"FeedsResponseSerializer - validation feed_type: '{feed_type}'")
-        return feed_type_validation(feed_type)
+@cache
+def feed_type_validation(feed_type: str) -> bool:
+    general_honeypots = GeneralHoneypot.objects.all().filter(active=True)
+    valid_types = set(["log4j", "cowrie", "all"] + [hp.name.lower() for hp in general_honeypots])
+    return feed_type in valid_types
```

Comment:
caching the query is a great idea, thank you!

However, this would make the output of the `GeneralHoneypot.objects.all().filter(active=True)` always the same, even if an admin changed something in that table. So this should not be cached in this way.

In my opinion, the better solution would be to calculate the items from GeneralHoneypots only once at the start of the available API views (`feeds` and `feeds_pagination`) and then pass the result as a parameter to the `get_queryset` and the `feeds_response` function and then, utlimately, to the Serializer for the usage.

Reply:
yep, I mean, I don't want to force any kind of specific pattern, like the DRF, because I think it is better to find a solution fast than to focus to code perfectionism because that can be always opinionable.

Another option can be using the [object-level validation](https://www.django-rest-framework.org/api-guide/serializers/#object-level-validation) and [extra-context](https://www.django-rest-framework.org/api-guide/serializers/#including-extra-context). In this way you can pass all the data that you want through the serializers and with the more generic validator you can do all the stuff that you want

Reply:
The [extra-context](https://www.django-rest-framework.org/api-guide/serializers/#including-extra-context) hint was what i needed! Thanks a lot! :)